File: iosApp/iosApp/App/AppDelegate/MultiConfigure.swift
//
// Created by Julius Canute on 15/5/20.
// Copyright (c) 2020 Julius Canute. All rights reserved.
//

import Foundation
import MultiConfigCommon
import UIKit

public class MultiConfigure {
    let starter: ConfigBase
    let dependencyInjectionContainer: ConfigurationDependencyContainer

    init(starter: ConfigBase) {
        self.starter = starter
        self.dependencyInjectionContainer = ConfigurationDependencyContainer(starter: starter)
    }

    public func getConfigurationManager() -> ConfigurationManager {
        starter.getConfigurationManager()
    }

    public func getConfig() -> ConfigurationGetter {
        starter.getConfig()
    }

    public func getLaunchController() -> LaunchController {
        starter.getLaunchController()
    }

    public func getEnvironment() -> String {
        starter.getEnvironment()
    }

    public func setEnvironment(environment: String) {
        starter.setEnvironment(environment: environment)
    }

    public func getRootViewController() -> UINavigationController {
        dependencyInjectionContainer.makeMainViewController()
    }
}

File: app/src/commonMain/kotlin/com/juliuscanute/multiconfig/config/MultiConfigure.kt
package com.juliuscanute.multiconfig.config

import com.juliuscanute.multiconfig.model.ConfigurationGetter
import com.juliuscanute.multiconfig.model.ConfigurationManager
import kotlin.native.concurrent.ThreadLocal

@ThreadLocal
object MultiConfigure {
    private lateinit var configBase: ConfigBase

    fun getConfigurationManager(): ConfigurationManager {
        return configBase.getConfigurationManager()
    }

    fun getConfig(): ConfigurationGetter {
        return configBase.getConfig()
    }

    fun getLaunchController(): LaunchController<*> {
        return configBase.getLaunchController()
    }

    fun getEnvironment(): String {
        return configBase.getEnvironment()
    }

    operator fun invoke(configBase: ConfigBase) {
        this.configBase = configBase
    }

    operator fun invoke(environment: String) {
        configBase.setEnvironment(environment = environment)
    }
}